test/studies/ml/lib/Chai.chpl

```diff
@@ -0,0 +1,828 @@
+        //     return backwardPropHelp(this.layers,this.layers.size - 1,x);
+        // }
+        // proc ref backwardProp(x: Tensor(?), delta: Tensor(?)) {
+        //     return backwardForwardPropHelp(this.layers,0,x,delta);
+        // }
+
+        proc ref forwardProp(input: Tensor(?)) {
+            const inputs = [input];
+            return this.forwardPropBatch(inputs)[0];
+        }
+
+        proc ref backwardProp(input: Tensor(?), delta: Tensor(?)) {
+            const inputs = [input];
+            const deltas = [delta];
+            return this.backwardPropBatch(inputs,deltas)[0];
+        }
+
+        proc ref forwardPropBatch(inputs) {
+            return forwardPropHelpBatch(this.layers, 0, inputs);
+        }
+        proc ref backwardPropBatch(inputs, deltas) {
+            return backwardForwardPropHelpBatch(this.layers,0,inputs=inputs,lastDeltas=deltas);
+        }
+
+        proc ref optimize(mag: real) {
+            for param i in 0..#(this.layers.size) {
+                this.layers[i].optimize(mag);
+            }
+        }
+        proc ref resetGradients() {
+            for param i in 0..#(this.layers.size) {
+                this.layers[i].resetGradients();
+            }
+        }
+
+        proc cost(x: Tensor(?), y: Tensor(?)) {
+            const z = this.forwardProp(x);
+            return tn.frobeniusNormPowTwo(y - z);
+        }
+
+        proc optimize(x: Tensor(?),y: Tensor(?),mag: real) {
+            const z = this.forwardProp(x);
+            const delta = z - y;
+            backwardForwardPropHelp(this.layers,0,x,delta);
+            // this.optimize(mag);
+        }
+
+
+
+        proc train(data, learningRate: real) {
+            var cost = 0.0;
+            forall ((x,y),i) in zip(data,0..) with (ref this, ref cost) {
+                // this.optimize(x,y,learningRate);
+                const z = this.forwardProp(x);
+                const delta = z - y;
+                cost += tn.frobeniusNormPowTwo(delta);
+                backwardForwardPropHelp(this.layers,0,x,delta);
+                if i % ((data.size / 100):int + 1) == 0 {
+                    // try! IO.stderr.write(">");
+                    write(">");
+                    try! IO.stdout.flush();
+                }
+
+            }
+            // try! IO.stderr.writeln();
+            writeln();
+            cost /= data.size;
+            writeln("Optimizing...");
+            for param i in 0..#(layers.size) {
+                layers[i].optimize(learningRate / data.size);
+            }
+
+            return cost;
+        }
+
+        proc save(path: string) throws {
+            var file = IO.open(path, IO.ioMode.cw);
+            var serializer = new IO.BinarySerializer(IO.ioendian.big);
+            var fw = file.writer(serializer=serializer);
+            // fw.write("[network]");
+            fw.write(layers.size);
+            for param i in 0..#(layers.size) {
+                layers[i].write(fw);
```

Comment:
Going off of the earlier comment about `proc write` vs. `proc writeThis`, this would change to:
```chapel
fw.write(layers[i]);
```

Or you could implement a `writeThis` for the whole network, allowing you to replace a bunch of this code with:
```chapel
file.writer(serializer=new IO.BinarySerializer(IO.ioendian.big)).write(this);
```

Reply:
Ok. I will commit the changes I have made for your other suggestions and start on this.

Reply:
Is it possible to differentiate the `writeThis` method in each layer, based on what kind of serializer is being used?

Reply:
Yeah, you can do something different for different serializers by defining a `serialize` method, and querying the filewriter's serializer type. See this example from the Linked List module: https://github.com/chapel-lang/chapel/blob/0ce54da1aeb6a40139c16e1ba2ed1a50230630d3/modules/packages/LinkedLists.chpl#L326-L335